### PR TITLE
[CLI UI kit - Themes adoption] Extract the ownership of development themes

### PR DESCRIPTION
**Pull request by Poitrin**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/internal-cli-foundations/issues/509
Fixes https://github.com/Shopify/cli/issues/1084
Fixes https://github.com/Shopify/cli/issues/1263

Requires https://github.com/Shopify/shopify-cli/pull/2724

### WHAT is this pull request doing?

* Generates Development Theme ID and calls Themes API in Node.js CLI.
* Depending on the `shopify theme` command: passes ID of generated theme to Ruby CLI for further processing.

### How to test your changes?

Make sure that your local Ruby CLI uses changes from https://github.com/Shopify/shopify-cli/pull/2724

1. Verify that `(pnpm run )shopify theme info` says `Development Theme ID   Not set`
2. Create a new development theme: `shopify theme dev --path="./my-theme"`
3. Verify that `shopify theme info` now displays the ID for the created dev theme
4. Verify that `shopify theme open -d` opens the dev theme (with the correct ID)
5. Verify that `shopify theme delete -d` deletes the dev theme (after you confirm the deletion)
6. Verify that `shopify theme info` says `Development Theme ID   Not set` again
7. Verify that `shopify theme pull -d --path="./my-theme"` throws the error `No development theme ID has been set. Please create a development theme first.`
8. Verify that `shopify theme push -d --path="./my-theme"` creates a new development theme.
9. Verify that `shopify theme info` now displays the ID for the created dev theme
10. Verify that `shopify theme dev --path="./my-theme"` uses the correct ID
11. Delete the theme with CLI 2.x (so that 3.x doesn't know anything about it): `shopify(-dev) theme delete <your ID>`
12. Verify that `shopify theme dev --path="./my-theme"` still works, but has created a new ID
13. Delete this theme also with CLI 2.x
14. Verify that `shopify theme pull -d --path="./my-theme"` throws the error `Development theme <old ID> could not be found. Please create a new development theme.`

#### `--theme-editor-sync`
1. `shopify theme init` a new theme.
2. Make sure that no development theme ID is set in CLI.
3. `shopify theme dev --store="…" --path="./draft" --theme-editor-sync`
4. Verify that CLI does not display any "Keep the remote version" prompt.

#### Keep Development Theme ID when switching stores
1. Execute `shopify theme dev --path="./my-theme" --store=<store 1>` and stop the server
2. Note down the theme ID that was created
3. Execute `shopify theme dev --path="./my-theme" --store=<store 2>` and stop the server
4. Note down the theme ID that was created
5. Execute `shopify theme dev --path="./my-theme" --store=<store 1>` again
6. Verify that the ID for store 1 was reused
7. Execute `shopify theme dev --path="./my-theme" --store=<store 2>` and stop the server
8. Verify that the ID for store 2 was reused

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/cli-kit/src/store.ts [hunk 1]
```diff
@@ -107,6 +107,21 @@ export function clearAllAppInfo(): void {
   store.clearAllAppInfo()
 }
 
+export function getDevelopmentTheme(): string | undefined {
+  const store = cliKitStore()
+  return store.getDevelopmentTheme()
+}
```
[the remark this answers is not in the document]
Comment (Poitrin) [reply packages/cli-kit/src/store.ts:113]: Awesome, thanks for the "warning" :-)

**Review by karreiro — COMMENTED** [review]
Thank you for the updates, @Poitrin! 🚀 

I've tried the PR and almost everything is working as expected. I've noticed a minor regression on this issue https://github.com/Shopify/shopify-cli/issues/2433 / https://github.com/Shopify/shopify-cli/pull/2463.

For fixing that scenario, I believe may rely on a hidden flag in the CL2 to differentiate themes created at runtime. Feel free to reach out if you need any extra context tho :)

Thanks again for this PR!

File: packages/cli-kit/src/store.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Poitrin) [reply packages/cli-kit/src/store.ts]: Right, thanks for highlighting this! I am using now `themeConf()` :+1: 😊

**Comment by Poitrin** [comment]
> I've noticed a minor regression on this issue [Shopify/shopify-cli#2433](https://github.com/Shopify/shopify-cli/issues/2433) / [Shopify/shopify-cli#2463](https://github.com/Shopify/shopify-cli/pull/2463).

Phew, good to have you as a reviewer, to think of all these sneaky regressions 😥.

> For fixing that scenario, I believe may rely on a hidden flag in the CL2 to differentiate themes created at runtime.

→ https://github.com/Shopify/shopify-cli/pull/2724

File: packages/theme/src/cli/commands/theme/dev.ts [hunk 2]
```diff
@@ -96,30 +98,40 @@ export default class Dev extends ThemeCommand {
...
   async run(): Promise<void> {
-    const {flags} = await this.parse(Dev)
+    let {flags} = await this.parse(Dev)
+    const store = ensureThemeStore(flags)
+    const adminSession = await ensureAuthenticatedThemes(store, flags.password, [], true)
+    const theme = await new DevelopmentThemeManager(adminSession).findOrCreate()
+    flags = {
+      ...flags,
+      theme,
+      'overwrite-json': Boolean(flags['theme-editor-sync']),
```
[the remark this answers is not in the document]
Comment (Poitrin) [reply packages/theme/src/cli/commands/theme/dev.ts]: It was in the end such a small change that I decided to include it in _this_ PR.
Moreover, as you suggested to make my changes "ready for host theme changes", I moved parts of them into this PR.

File: .eslintrc.cjs [hunk 3]
```diff
@@ -188,6 +188,7 @@ module.exports = {
         '**/public/node/plugins/tunnel.ts',
         '**/public/node/presets.ts',
         '**/public/node/result.ts',
+        '**/public/node/themes/**/*',
```
Comment (Poitrin) [inline .eslintrc.cjs:191]: → https://github.com/Shopify/cli/pull/1264#discussion_r1094585508

**Review by matteodepalo — APPROVED** [review]
LGTM

**Pull request merged**